### PR TITLE
ai_demos

### PR DESCRIPTION
**Pull request by silkspace**
… rather than just those with found edges in text_utils.py, and adds what should be on master in feature_utiils

### CODE REVIEW

**Review by lmeyerov — CHANGES_REQUESTED** [review]
It'll take me some time to work through the ipynb, but a separate PR on the util changes (if helpful) looks fine...

File: graphistry/feature_utils.py [hunk 1]
```diff
@@ -2608,3 +2648,76 @@ def _featurize_or_get_edges_dataframe_if_X_is_None(
             reuse_if_existing=True,
             memoize=memoize,
         )
+        
+    def get_matrix(self, column, kind='nodes'):
```
Comment (lmeyerov) [inline graphistry/feature_utils.py:2652]: let's give it a name that makes sense to users scrolling through, like `.features_by_cols(cols=[...])`

File: graphistry/feature_utils.py [hunk 2]
```diff
@@ -2608,3 +2648,76 @@ def _featurize_or_get_edges_dataframe_if_X_is_None(
...
+        condprobs = conditional_probability(x, given, df)
+        
+        cprob = pd.DataFrame(list(condprobs.index), columns=[given, x])
+        cprob['_probs'] = condprobs.values
+    
+        res = res.edges(cprob, x, given).bind(edge_weight='_probs', *args, **kwargs)
+        
+        return res
+    
+    def conditional_probs(self, x, given, kind='nodes', how='index'):
```
Comment (lmeyerov) [inline graphistry/feature_utils.py:2702]: maybe belongs in compute, not features?

File: graphistry/feature_utils.py (hunk 1 above)
Comment (silkspace) [reply graphistry/feature_utils.py]: changed this to `get_features_by_cols(self, columns: Union[List, str], kind: str = 'nodes')`

File: graphistry/feature_utils.py (hunk 2 above)
Comment (silkspace) [reply graphistry/feature_utils.py]: added new class mixing in .compute

File: demos/ai/Introduction/Ask-HackerNews-Demo.ipynb [hunk 3]
```diff
@@ -0,0 +1,569 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "id": "c39da4a9",
+   "metadata": {},
+   "source": [
+    "# Hello Graphistry[ai]\n",
+    "\n",
+    "`Graphistry[ai]` removes 80% of the work of cleaning, connecting, and encoding data, while increasing `Time to Graph` and value 100 fold in as little as two lines of code.\n",
```
Comment (lmeyerov) [inline demos/ai/Introduction/Ask-HackerNews-Demo.ipynb:10]: PyGraphistry[ai] here an eslewhere?

Comment (silkspace) [reply demos/ai/Introduction/Ask-HackerNews-Demo.ipynb]: Thanks, it's been changed.

**Comment by silkspace** [comment]
Should be good to go.

File: graphistry/feature_utils.py [hunk 4]
```diff
@@ -220,7 +220,7 @@ def safe_divide(a, b):
 
 
 def features_without_target(
-    df: pd.DataFrame, y: Optional[Union[List[str], str, pd.DataFrame]] = None
+    df: pd.DataFrame, y: Optional[Union[List, str, pd.DataFrame]] = None
```
Comment (lmeyerov) [inline graphistry/feature_utils.py:223]: mypy failing line 517: https://github.com/graphistry/pygraphistry/actions/runs/3734996664/jobs/6337709428

File: graphistry/compute/conditional.py [hunk 5]
```diff
@@ -0,0 +1,118 @@
+import logging
+import pandas as pd
+from typing import Any, List, Union, TYPE_CHECKING
+from typing_extensions import Literal
+
+from graphistry.Engine import Engine
+from graphistry.Plottable import Plottable
+
+logger = logging.getLogger("compute.conditional")
```
Comment (lmeyerov) [inline graphistry/compute/conditional.py:9]: common pattern here is `logging.getLogger(__file__)`, but nbd

File: graphistry/compute/conditional.py [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def conditional_graph(self, x, given, kind='nodes', *args, **kwargs):
+        """
+        conditional_graph -- p(x|given) = p(x, given) / p(given)
+        
+        Useful for finding the conditional probability of a node or edge attribute
+        
+        returned dataframe sums to 1 on each column
```
Comment (lmeyerov) [inline graphistry/compute/conditional.py:68]: for public methods expected to be used by non-core-team, helps to add examples as well. this is such a cool method so thought i'd comment.

non-blocker.

File: graphistry/compute/conditional.py [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    res = pd.crosstab(df[x], df[given], margins=True, normalize=how)
+    if how == 'index':  # normalize over columns so .sum(0) = 1 irrespective of `how`
+        return res.T
+    return res
+
+class ConditionalMixin(MIXIN_BASE):
+    def __init__(self, *args, **kwargs):
+        pass
+
+    def conditional_graph(self, x, given, kind='nodes', *args, **kwargs):
```
Comment (lmeyerov) [inline graphistry/compute/conditional.py:62]: for another day, as the intent here is often to help plot big graphs, would be cool if we could also include some sort of % sampling or thresholding

File: graphistry/dgl_utils.py [hunk 8]
```diff
@@ -555,156 +531,155 @@ def __getitem__(self, idx):
...
-    # # trained comparison
-    # logits = model(G, node_features)
-    # pred = logits.argmax(1)
-    #
-    # accuracy = sum(pred[test_mask] == labels[test_mask]) / len(
-    #     pred[test_mask]
-    # )  # does pretty well!
-    # print("-" * 60)
-    # print(f"Final Accuracy: {100 * accuracy:.2f}%")
+#     import graphistry
```
Comment (lmeyerov) [inline graphistry/dgl_utils.py:534]: could we just delete?

or this helps dev somehow, and we might as well uncomment?

File: graphistry/feature_utils.py [hunk 9]
```diff
@@ -2606,3 +2599,40 @@ def _featurize_or_get_edges_dataframe_if_X_is_None(
...
+            X = self._node_features
+        elif kind == 'edges' and hasattr(self, '_edge_features'):
+            X = self._edge_features
+        else:
+            raise ValueError('make sure to call `featurize` or `umap` before calling `get_features_by_cols`')
+        
+        transformed_columns = X.columns[X.columns.map(lambda x: True if column_part in x else False)]  # type: ignore
+        return X[transformed_columns]  # type: ignore
+    
+    def get_features_by_cols(self, columns: Union[List, str], kind: str = 'nodes'):
```
Comment (lmeyerov) [inline graphistry/feature_utils.py:2614]: Maybe helpful for the docs, if I wanted all features generated for input column `"src_ip"`, could I use this? (Are original column names used for the prefix of generated feature column names?)

Either way, maybe useful to clarify here. 

I keep wondering every time I see this helper.. how to understand feature provenance..

File: graphistry/features.py [hunk 10]
```diff
@@ -0,0 +1,241 @@
+from collections import UserDict
+from .util import setup_logger
+from .constants import VERBOSE, TRACE
+
+logger = setup_logger("graphistry.features", verbose=VERBOSE, fullpath=TRACE)
```
Comment (lmeyerov) [inline graphistry/features.py:5]: see comment about `__file__`

File: graphistry/outliers.py [hunk 11]
```diff
@@ -0,0 +1,173 @@
+from typing import Any, List, Optional, Union
+
+import matplotlib.font_manager
+import matplotlib.pyplot as plt
```
Comment (lmeyerov) [inline graphistry/outliers.py:4]: are methods in `outliers.py` lazily imported or always-on top-level imports? if not, defer import to use b/c matplotlib dependency is optional

probably less fragile if we make just-in-time imports either way..

File: graphistry/outliers.py [hunk 12]
```diff
@@ -0,0 +1,173 @@
...
+import pandas as pd
+import numpy as np
+
+from sklearn import neighbors
+from sklearn.covariance import EllipticEnvelope
+from sklearn.svm import OneClassSVM
```
Comment (lmeyerov) [inline graphistry/outliers.py:10]: we should probably defer these dependencies too

File: graphistry/text_utils.py [hunk 13]
```diff
@@ -270,6 +271,7 @@ def search_graph(
 
     def save_search_instance(self, savepath):
         from joblib import dump  # type: ignore   # need to make this onnx or similar
+        self.build_index()
```
Comment (lmeyerov) [inline graphistry/text_utils.py:274]: maybe we make `self.search_index` a lazy auto-memoizing getter? i'm not sure when it invalidates, but the peppering of `build_index()` sounds like we're doing that anyway, so maybe centralize

**Review by lmeyerov — CHANGES_REQUESTED** [review]
I reviewed all non-`.ipynb`:

* [ ] mypy failing again, those pandas/mypy people...

* [ ] `.py` files look pretty good! check out inline comments if anything interesting but afaict fine. thinks for removing most of the `noqa`s.
* i tested import time and seems good, most of it is just `pandas`, and torch seems to be getting correctly deferred
* once green, let's do an initial landing & version bump, and i'll try to do a fast-follow on editing the ipynb's before we advertise

**Pull request closed without merging**